File: src/ShieldBleuette/Ino/lib/ServoController/servocontroller.cpp
/*
    ServoController - Bleuette servo controller
    Copyright (c) 2013 Charles Rincheval.  All rights reserved.

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

    The latest version of this library can always be found at
    http://www.digitalspirit.org/
*/

#include "servocontroller.h"

ServoController::ServoController() :
    _serial(Serial1)
{
    _last_status_code = 0;
    memset(&_buffer, 0x00, sizeof(_buffer));
    memset(&_current_values, 0x00, sizeof(_current_values));

    _serial.begin(9600);
}

ServoController::ServoController(HardwareSerial &serial) :
    _serial(serial)
{
    _last_status_code = 0;

    memset(&_buffer, 0x00, sizeof(_buffer));
    memset(&_current_values, 0x00, sizeof(_current_values));
};

inline bool ServoController::getLastStatus()
{
    return (_last_status_code == SERVO_ACK) ? true : false;
}

unsigned long ServoController::get(unsigned char servo)
{
    unsigned long servos[] = {
        SERVO_0,
        SERVO_1,
        SERVO_2,
        SERVO_3,
        SERVO_4,
        SERVO_5,
        SERVO_6,
        SERVO_7,
        SERVO_8,
        SERVO_9,
        SERVO_10,
        SERVO_11,
        SERVO_12,
        SERVO_13
    };

    return (servo < sizeof(servos) ? servos[servo] : 0);
}

int ServoController::getResponse()
{
    while (!_serial.available());
    return _serial.read();
}

inline void ServoController::write(unsigned char byte)
{
    _serial.write(byte);
    delay(SERVO_DELAY_BYTE);
}

inline bool ServoController::command(SERVO_CMD command)
{
    write(SERVO_HEADER);
    delay(SERVO_DELAY_BYTE);
    write(command);
    delay(SERVO_DELAY_BYTE);

    _last_status_code = getResponse();

    return getLastStatus();
}

bool ServoController::command(SERVO_CMD command, unsigned char data[], unsigned char size)
{
    unsigned char c = 0;
    unsigned char control = 0;
    unsigned char retry = SERVO_MAX_RETRY;

    do {
        write(SERVO_HEADER);
        write(command);

        control = 0;
        for (c = 0; c < size; c++) {
#if SERVO_DEBUG
            PLN(data[c]);
#endif
            write(data[c]);

            control += data[c];
        }

        write(control);
        delay(SERVO_DELAY_BYTE);

        _last_status_code = getResponse();
        if (_last_status_code == SERVO_ACK) {
            //PLN("MEMCPY!");
            memcpy(&_current_values, data, SERVO_COUNT);
            break;
        }

    } while (retry--);

    return getLastStatus();
}

bool ServoController::init()
{
    return command(SERVO_CMD_INIT);
}

/**
 *  Stop impulsion generation
 */
bool ServoController::pause()
{
    return command(SERVO_CMD_PAUSE);
}

/**
 *  Restart impulsion
 */
bool ServoController::resume()
{
    return command(SERVO_CMD_RESUME);
}

/**
 *  Set to 0 all consigne
 */
bool ServoController::clear()
{
    int status = command(SERVO_CMD_CLEAR);
    if (status == SERVO_ACK) {
        memset(&_current_values, 0x00, sizeof(_current_values));
    }
}

void ServoController::setValue(unsigned char index, unsigned char value)
{
    _buffer[index] = value;
}

unsigned char ServoController::getPosition(unsigned char servo) {
    if (servo >= 0 && servo <= SERVO_COUNT) {
        return _current_values[servo];
    }
}

void ServoController::getPositions(unsigned char &buf) {
    memcpy(&buf, _current_values, SERVO_COUNT);
}

/**
 *  Set servo values
 *  servos is a bits field and values is the values only for bit set
 *
 *  Example:
 *  - Set all servo to 200 :
 *      servos : 0xFFFF, values[] = 200, 200, 200, 200, ...
 *  - Set servos 0 & 2 to 100 :
 *      servos : 0b00000000 00000101, values[] = 100, 100
 *
 */
void ServoController::setValues(unsigned long servos, unsigned char * values, unsigned char count)
{
    unsigned char b, v = 0;
    for (b = 0, v = 0; v < count; b++) {
        if (CHECK_BIT(servos, b)) {
            // Todo: replace with setValue
            //_buffer[b] = values[v++];
            setValue(b, values[v++]);
        }
    }
}

void ServoController::sendValues()
{
    command(SERVO_CMD_SET, _buffer, sizeof(_buffer));
}

File: src/BleuettePi/Python/ServoController.py
#!/usr/bin/python

import serial
from time import sleep
from array import array

SERVO0 = 1
SERVO1 = 2
SERVO2 = 4
SERVO3 = 8
SERVO4 = 16
SERVO5 = 32
SERVO6 = 64
SERVO7 = 128
SERVO8 = 256
SERVO9 = 512
SERVO10 = 1024
SERVO11 = 2048
SERVO12 = 4096
SERVO13 = 8192

ANA0 = '0'
ANA1 = '1'
ANA2 = '2'
ANA3 = '3'
ANA4 = '4'
ANA5 = '5'

class ServoController:

    SERVO_COUNT = 14

    HEADER = chr(255)

    CMD_INIT    = 'I'   # Init
    CMD_PAUSE   = 'P'   # Pause all servo
    CMD_RESUME  = 'R'   # Resume from pause
    CMD_CLEAR   = 'C'   # Clear all values
    CMD_SET     = 'S'   # Set consigne

    CMD_READAN0 = '0'   # Read analog value
    CMD_READAN1 = '1'   # Read analog value
    CMD_READAN2 = '2'   # Read analog value
    CMD_READAN3 = '3'   # Read analog value
    CMD_READAN4 = '4'   # Read analog value
    CMD_READAN5 = '5'   # Read analog value

    DELAY_BYTE = 0.002

    ACK = 'O'
    NACK = 'N'

    last_status_code = ''

    values = array('c', [chr(200)] * SERVO_COUNT)

    def __init__(self, device):
        self.com = serial.Serial(device, 9600, timeout = 1)
        #ser.open()

    def getResponse(self):
        return self.com.read(1)

    def getLastStatus(self):
        return (self.last_status_code == self.ACK)

    # Send only one command
    def command(self, cmd):
        self.com.write(self.HEADER)
        sleep(self.DELAY_BYTE)
        self.com.write(cmd)
        self.last_status_code = self.getResponse()
        return self. getLastStatus()

    def sendValues(self):
        self.com.write(self.HEADER)
        sleep(self.DELAY_BYTE)
        self.com.write(self.CMD_SET)

        control = 0
        for i in range(0, self.SERVO_COUNT):
            self.com.write(self.values[i])

            control += ord(self.values[i])

        #sleep(self.DELAY_BYTE)
        #print control % 255
        #print control % 256
        #self.com.write(chr(78)) #chr(control % 255))
        self.com.write(chr(control % 256))
        
        #sleep(self.DELAY_BYTE)

        self.last_status_code = self.getResponse()
        return self.getLastStatus()

    def init(self):
        return self.command(self.CMD_INIT)

    # Pause
    def pause(self):
        return self.command(self.CMD_PAUSE)

    # Resume
    def resume(self):
        return self.command(self.CMD_RESUME)

    # Set to 0 all consigne
    def clear(self):
        return self.command(self.CMD_CLEAR)

    # Read analog value
    def readAnalog(self, index):
        self.com.write(self.HEADER)
        sleep(self.DELAY_BYTE)
        self.com.write(index)

        #print index, ord(index), chr(index)

        #self.last_status_code = self.getResponse()
        return self.getResponse()
        #return self.getLastStatus()

    def setValue(self, index, value):
        values[index] = value 

'''
    void setValue(unsigned char, unsigned char);

    unsigned char getPosition(unsigned char);
    void getPositions(unsigned char &);

    void setValues(unsigned long, unsigned char *, unsigned char);
    void sendValues();

    unsigned long get(unsigned char);

    bool command(SERVO_CMD, unsigned char [], unsigned char);
'''

if __name__ == '__main__':

    Servo = ServoController('/dev/ttyAMA0')
    print Servo.init()
    print Servo.sendValues()
    print Servo.resume()
    print Servo.readAnalog(ANA2)

